### PR TITLE
Mulcat option

### PR DESCRIPTION
**Pull request by JunzheJosephZhu**

### CODE REVIEW

**Comment by mpariente** [comment]
Great this time, thanks !

**Comment by mpariente** [comment]
@JorisCos do you mind reviewing please?

File: asteroid/masknn/recurrent.py [hunk 1]
```diff
@@ -196,22 +258,31 @@ def __init__(
...
+            )
         # IntraRNN and linear projection layer (always bi-directional)
-        self.intra_RNN = SingleRNN(
-            rnn_type, in_chan, hid_size, num_layers, dropout=dropout, bidirectional=True
-        )
-        self.intra_linear = nn.Linear(hid_size * 2, in_chan)
+        self.intra_linear = nn.Linear(self.intra_RNN.output_size, in_chan)
         self.intra_norm = norms.get(norm_type)(in_chan)
+
         # InterRNN block and linear projection layer (uni or bi-directional)
```
Comment (JorisCos) [inline asteroid/masknn/recurrent.py:293]: Same here

File: asteroid/masknn/recurrent.py [hunk 2]
```diff
@@ -196,22 +258,31 @@ def __init__(
...
+            rnn_type, in_chan, hid_size, num_layers, dropout=dropout, bidirectional=bidirectional
+            )
+        else:
+            self.intra_RNN = SingleRNN(
+                rnn_type, in_chan, hid_size, num_layers, dropout=dropout, bidirectional=True
+            )
+            self.inter_RNN = SingleRNN(
+            rnn_type, in_chan, hid_size, num_layers, dropout=dropout, bidirectional=bidirectional
+            )
         # IntraRNN and linear projection layer (always bi-directional)
```
Comment (JorisCos) [inline asteroid/masknn/recurrent.py:289]: This comment should be before `  self.intra_RNN = DoubleRNN(...)`

**Review by JorisCos — COMMENTED** [review]
Thanks for the PR, this looks really nice. A minor fix on the docs is needed otherwise LGTM

File: asteroid/masknn/recurrent.py [hunk 3]
```diff
@@ -62,6 +62,67 @@ def forward(self, inp):
         return rnn_output
 
 
+class DoubleRNN(nn.Module):
+    """Module for a RNN block.
+
+    Inspired from https://github.com/yluo42/TAC/blob/master/utility/models.py
+    Licensed under CC BY-NC-SA 3.0 US.
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:69]:
```suggestion
```

Comment (mpariente) [reply asteroid/masknn/recurrent.py:69]: I don't think it's inspired by this repo right?

File: asteroid/masknn/recurrent.py [hunk 4]
```diff
@@ -62,6 +62,67 @@ def forward(self, inp):
         return rnn_output
 
 
+class DoubleRNN(nn.Module):
+    """Module for a RNN block.
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:66]: Need to be more explicit.

**Review by mpariente — COMMENTED** [review]
Let's think about a way which doesn't copy paste the `SingleRNN` code in `DoubleRNN`.

**Comment by mpariente** [comment]
@JunzheJosephZhu ?

**Comment by JunzheJosephZhu** [comment]
I guess the best way to do it would be combining them into s single object, adding an option for single/doubleRNN and adding a switch statement into the code?

For example,
class singleRNN(...):
   def init(...,...,mulcat=False)
   If mulcat:
       self.inter_RNN = ...
   else:
       self.inter_RNN1 =
       self.inter_RNN2 =

The code would be somewhat more confusing to readers though

File: asteroid/masknn/recurrent.py [hunk 5]
```diff
@@ -62,6 +62,64 @@ def forward(self, inp):
...
+        return self.hidden_size * (2 if self.bidirectional else 1) + self.input_size
+
+    def forward(self, inp):
+        """ Input shape [batch, seq, feats] """
+        self.rnn1.flatten_parameters()  # Enables faster multi-GPU training.
+        self.rnn2.flatten_parameters()  # Enables faster multi-GPU training.
+        output = inp
+        rnn_output1, _ = self.rnn1(output)
+        rnn_output2, _ = self.rnn2(output)
+        return torch.cat((rnn_output1 * rnn_output2, output), -1)
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:120]:
```suggestion
        rnn_output1, _ = self.rnn1(inp)
        rnn_output2, _ = self.rnn2(inp)
        return torch.cat((rnn_output1 * rnn_output2, inp), -1)
```

**Comment by mpariente** [comment]
Let's keep the code like this for now. 
Please
- Rename `DoubleRNN` to `MulCatRNN`
- Add a description in the docstring. 
- Test the `mul_cat` option in jit_test.py

**Comment by JunzheJosephZhu** [comment]
It looks like when running the jit test, sudormrf always fails the test. Is this something I need to worry about, or is this unrelated to DPRNN?

File: asteroid/masknn/recurrent.py [hunk 6]
```diff
@@ -62,6 +62,63 @@ def forward(self, inp):
         return rnn_output
 
 
+class MulCatRNN(nn.Module):
+    """Module for a MulCat RNN block.
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:66]:
```suggestion
    """MulCat RNN block from [1].
    
    Composed of two RNNs, returns ``cat([RNN_1(x) * RNN_2(x), x])``.
```

Comment (mpariente) [reply asteroid/masknn/recurrent.py:66]: Need to add reference, as done in other models.

File: tests/jit/jit_models_test.py [hunk 7]
```diff
@@ -106,7 +107,7 @@ def small_model_params():
     return params
 
 
-@pytest.mark.parametrize("model_def", [DCCRNet, DeMask])
+@pytest.mark.parametrize("model_def", [DPRNNTasNet])
```
Comment (mpariente) [inline tests/jit/jit_models_test.py:110]: That doesn't look right. Why did you do that?

**Review by mpariente — COMMENTED** [review]
We are getting much closer, thanks ! What's left

- [ ] Add `use_mulcat` is `get_config` so that serialization will work
- [ ] Add the `use_mulcat` parameterize in `tests/models/models_test.py` to ensure serialization will work. 
- [ ] Fix the jit tests which don't look right to me. 
- [ ] Add the reference to the original MulCat paper in `MulCatRNN`

File: tests/jit/jit_models_test.py (hunk 7 above)
Comment (JunzheJosephZhu) [reply tests/jit/jit_models_test.py:110]: I think we just add whatever model we just want to test with jit here, no?

**Comment by JunzheJosephZhu** [comment]
Also, do you want to implement multi-stage loss? That was also in the facebook paper.

File: tests/jit/jit_models_test.py (hunk 7 above)
Comment (mpariente) [reply tests/jit/jit_models_test.py]: We need to keep all the tests, otherwise we don't know when we break things.

Comment (JunzheJosephZhu) [reply tests/jit/jit_models_test.py]: I added them back

**Comment by mpariente** [comment]
Eventually, yes, but in the recipe sounds fine. 

> since I'm gonna re-write the backbone anyway.

What do you mean?

**Comment by JunzheJosephZhu** [comment]
I mean I will use DPRNNBlock but not DPRNN for my separator, since the part starting from Conv2D diverge.

**Comment by mpariente** [comment]
Alright. 
Can you fix the other issues I mentioned so that we can merge?

**Comment by JunzheJosephZhu** [comment]
Ok, I'll do it ASAP. Btw, how do you write test cases for dataset objects?

**Comment by mpariente** [comment]
We don't have tests for datasets.. Which is a large limitation.

**Comment by JunzheJosephZhu** [comment]
I just finished my work with variable speaker wsj0mix dataset in my recipe folder. Should I start a PR yet?

File: tests/models/models_test.py [hunk 8]
```diff
@@ -109,7 +109,9 @@ def test_save_and_load_convtasnet(fb, sample_rate):
 
 
 def test_dprnntasnet_sep():
-    nnet = DPRNNTasNet(n_src=2, n_repeats=2, bn_chan=16, hid_size=4, chunk_size=20, n_filters=32)
+    nnet = DPRNNTasNet(
+        n_src=2, n_repeats=2, bn_chan=16, hid_size=4, chunk_size=20, n_filters=32, use_mulcat=True
```
Comment (mpariente) [inline tests/models/models_test.py:113]: It's not what I meant. 
I mean to use the `@pytest.mark.parameterize` to test both. 
I'll take care of it.

**Comment by JunzheJosephZhu** [comment]
I have no clue why it's not passing the test_save_and_load_dprnn. Do you have any idea?

**Comment by mpariente** [comment]
The loading is failing. 

We need to test the mulcat version, but still keep the old tests. For that, we can use `@pytest.mark.parameterize`. 

Regarding the recipe, yes, you can make a PR :smiley:  Thanks

**Comment by mpariente** [comment]
Because `use_mulcat` is not returned in `get_config` (as I mentioned before) so the new model doesn't use it.

**Comment by JunzheJosephZhu** [comment]
Where do I find get_config?

**Comment by JunzheJosephZhu** [comment]
Hi, I'm starting to write the model class. I noticed that in all your models that are in the source code just use make_optimizer, but those in the recipe define make_model_and_optimizer. Do I have to follow this?

**Comment by mpariente** [comment]
No, you don't have to. 
The most important is that loading the model for eval is easy.

**Comment by JunzheJosephZhu** [comment]
how do i write test cases for models in my recipe?
Also, github send me an email everyday telling me the version consistency check failed. I have no clue what it means

**Comment by mpariente** [comment]
> how do i write test cases for models in my recipe?

They aren't tested if they are in the recipe folder. 

> Also, github send me an email everyday telling me the version consistency check failed. I have no clue what it means

Sorry about that, I don't receive those e-mails.. Anybody else receives those e-mails?

File: asteroid/masknn/recurrent.py [hunk 9]
```diff
@@ -196,22 +257,51 @@ def __init__(
...
+                bidirectional=bidirectional,
+            )
+        else:
+            self.intra_RNN = SingleRNN(
+                rnn_type,
+                in_chan,
+                hid_size,
+                num_layers,
+                dropout=dropout,
+                bidirectional=bidirectional,
```
Comment (mpariente) [inline asteroid/masknn/recurrent.py:291]:
```suggestion
                 bidirectional=True,
```

**Comment by mpariente** [comment]
I
- Fixed the jit tests
- Parameterize model test with `use_mulcat`
- Restore always bidirectional IntraRNN
- Removed the recipe files (which should be added to another branch). 

This is ready for merging on my side. 
Next time, don't make PRs from your master branch please :wink:

**Comment by JunzheJosephZhu** [comment]
---------- Forwarded message ---------

**Comment by mpariente** [comment]
Not sure..
